Edit profile page

SheepTester opened the pull request:
<!-- Fill in all spots surrounded by brackets and confirm all check boxes after confirming completion of the tasks -->

# Info

Closes #87 

# Description

What changes did you make? List all distinct problems that this PR addresses. Explain any relevant
motivation or context.

Created the edit profile page, including its functionality.

I've also added some new components and utility functions, which would be helpful for other parts of the website as well:

- An accessible `Modal` component for general modals.
- A `Cropper` component for cropping user-selected images to a specific aspect ratio. It will also help discourage users from uploading non-images, and it enforces max file sizes by compressing the file as a JPG until it's small enough.
- A custom hook `useObjectUrl` that creates an object URL for the given `File`, and revokes it when it's done.
- An SCSS `accessibly-hidden` mixin for elements that should be tab-accessible but should also pretend to be hidden.

## Changes

- Created edit profile page
- Applied the new utility stuff above to other parts of the site as needed, including adding the cropper to the create event form.

# Type of Change

- [ ] Bug Fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as
      expected)
- [ ] Logistics Change (A change to a README, description, or dev workflow setup like
      linting/formatting)
- [ ] Continuous Integration Change (Related to deployment steps or continuous integration
      workflows)
- [ ] Other: (Fill In) <!-- Edit this type of change if you select this -->

# Testing

I have tested that my changes fully resolve the linked issue ...

- [x] locally on Desktop.
- [x] locally on mobile - use https://ngrok.io to get a copy on a mobile device
- [x] on the live deployment preview on Desktop.
- [x] on the live deployment preview on Mobile.
- [ ] I have run and passed all new and existing Cypress tests. Add screenshots below.

# Checklist

- [x] I have performed a self-review of my own code.
- [x] I have followed the style guidelines of this project.
- [x] I have documented my code's `src/lib` functions and commented hard to understand areas
      anywhere else.
- [x] My changes produce no new warnings.

# Screenshots

Please include a screenshot of your Cypress testing suite passing successfully.

If you made any visual changes to the website, please include relevant screenshots below.

Desktop, dark mode | Mobile, light mode
--- | ---
![OVERWRITE ME](https://github.com/acmucsd/membership-portal-ui-v2/assets/22133785/5f688cd6-0f4f-4128-8aa0-9acdc7db4dde) | ![OVERWRITE ME](https://github.com/acmucsd/membership-portal-ui-v2/assets/22133785/66f1b5e1-2943-4f53-9c1f-24af074e5998)

![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/22133785/535a81ea-8217-4a19-8f4a-5f1a715be2c3)

SheepTester committed:
EditField

SheepTester committed:
Add the other text fields

SheepTester committed:
Add socil media icons

SheepTester committed:
Adjust social media icon size to my preference

SheepTester committed:
Toggle switch, resume and user APIs

SheepTester committed:
Merge branch 'main' into sean/edit-profile

SheepTester committed:
enfatten the toggle

alex feedback

SheepTester committed:
Upload and delete resume functionality

SheepTester committed:
Merge branch 'sean/edit-profile' of https://github.com/acmucsd/membership-portal-ui-v2 into sean/edit-profile

SheepTester committed:
Button designs and save profile changes

SheepTester committed:
Upload profile photo

Also move Modal to common component

SheepTester committed:
Style the profile photo selector

SheepTester committed:
Image cropper

SheepTester committed:
Crop image with canvas

SheepTester committed:
Drag and drop files

SheepTester committed:
Reject non-images

farisashai committed:
Merge branch 'main' into sean/edit-profile

SheepTester committed:
Create custom hook `useObjectUrl`

could be useful for the Add to Apple calendar button

SheepTester committed:
Correctly scale image

SheepTester committed:
Mobile design and "Current profile" preview

SheepTester committed:
Merge remote-tracking branch 'origin/main' into sean/edit-profile

SheepTester committed:
Merge branch 'sean/edit-profile' of https://github.com/acmucsd/membership-portal-ui-v2 into sean/edit-profile

SheepTester committed:
Preview design, accordion heading design for mobile

SheepTester committed:
fix pfp size in preview

SheepTester committed:
Warn if unsaved changes, finish cropper design

SheepTester committed:
Avoid signing the user out when they update their profile

SheepTester committed:
Use `Modal` and `useObjectUrl` for the event cards

SheepTester committed:
Update API types

there is apparently support for social media and attendance history now

SheepTester committed:
Use Cropper for event management and autocompress cover image as needed

SheepTester committed:
Set social media URLs and isAttendancePublic

SheepTester committed:
Automatically fix non-URLs for social media

SheepTester committed:
Add other supported social media and show them in preview

alexzhang1618 commented on src/components/admin/event/EventDetailsForm/index.tsx:
```diff
@@ -267,24 +277,48 @@ const EventDetailsForm = (props: IProps) => {
+              file={selectedCover}
+              aspectRatio={1920 / 1080}
+              maxFileHeight={1080}
+              maxSize={config.file.MAX_EVENT_COVER_SIZE_KB * 1024}
```
I like this maxSize thing here as a warning for users!

one thing I think that's worth mentioning is that the max file size can vary between testing.api and prod (for testing it's 256 KB, for prod its 2048 KB), so on testing you might get some weird behavior:
- you try to upload an image that's in between 256 KB and 2048 KB
- the event is created successfully but the image doesn't get uploaded

this isn't a massive issue since it only affects testing and not prod, but it could be worth looking into having this config value switch depending on the deploy/runtime var

alexzhang1618 replied:
just for reference:

on prod: 
- event cover: 2048
- pfp: 256
- banner (don't think we use these currently): 2048
- merchPhoto: 1024
- resume: 2048

on testing/staging:
- event cover: 256
- pfp: 256
- banner (don't think we use these currently): 256
- merchPhoto: 1024
- resume: wasn't specified in caprover for some reason

SheepTester replied:
This is handled by the config file which sets the max size depending on prod vs dev

Regardless, the cropper is guaranteed to close with an image file under the specified file size. If the image is too large, it'll try compressing it with lower and lower quality until it fits under the maximum

alexzhang1618 replied:
oh hm not sure why I was having that bug earlier then. Do we want to not make the updateEvent API call if the image is too big?

SheepTester replied:
hmm what image were you using?

we could add a check in `AdminEventManager.createNewEvent` to prevent creating an event if the cover is too large, but `Cropper` should make the same check already

alexzhang1618 replied:
Does `Cropper` block form submission if the cover is too large? if not, then we should probably enforce a check here before allowing submission

SheepTester replied:
> Does `Cropper` block form submission if the cover is too large? if not, then we should probably enforce a check here before allowing submission

yes, but I can add another check inside `createNewEvent` as a sanity check just in case

alexzhang1618 commented on src/components/admin/event/EventDetailsForm/index.tsx:
```diff
@@ -267,24 +277,48 @@ const EventDetailsForm = (props: IProps) => {
+                  if (file) {
+                    setSelectedCover(file);
+                  }
+                }}
```
Why is this `onChange` async? at a glance the function looks synchronous

alexzhang1618 commented on src/components/common/Cropper/index.tsx:
```diff
@@ -0,0 +1,199 @@
+  // eslint-disable-next-line no-unused-vars
+  onCrop: (file: Blob) => void;
+  // eslint-disable-next-line no-unused-vars
+  onClose: (reason: 'invalid-image' | 'cannot-compress' | null) => void;
```
are these rules still supposed to be disabled here?

SheepTester replied:
Yeah, for some reason ESLint thinks the parameters are unused even though they're just part of the type signature. You can see the rule being disabled throughout the codebase

alexzhang1618 commented on src/components/common/Cropper/index.tsx:
```diff
@@ -0,0 +1,199 @@
+              onCrop(blob);
+              return;
+            }
+            // Try compressing as JPG with various qualities, in parallel due to
+            // eslint(no-await-in-loop)
+            const blobs = await Promise.all(
+              [1, 0.9, 0.7, 0.4, 0].map(quality => toBlob(canvas, 'image/jpeg', quality))
+            );
+            const firstSmallEnough = blobs.find(blob => blob && blob.size <= maxSize);
+            if (firstSmallEnough) {
+              onCrop(firstSmallEnough);
+            } else {
+              onClose('cannot-compress');
+            }
+          }}
```
🔥 🔥 🔥 this is epic

alexzhang1618 commented on src/components/profile/EditField/index.tsx:
```diff
@@ -0,0 +1,123 @@
+  // eslint-disable-next-line no-unused-vars
+  onChange: (value: string) => void;
+  // eslint-disable-next-line no-unused-vars
+  onBlur?: (value: string) => string;
```
same here, can these `eslint-disable`s be removed?

alexzhang1618 commented on src/components/profile/Preview/index.tsx:
```diff
@@ -0,0 +1,104 @@
+    default:
+      return null;
+  }
+}
```
We might want to move this out into a separate component since I think this'll be useful for the actual profile page

alexzhang1618 commented on src/components/profile/Switch/index.tsx:
```diff
@@ -0,0 +1,27 @@
+interface SwitchProps {
+  checked: boolean;
+  // eslint-disable-next-line no-unused-vars
+  onCheck: (checked: boolean) => void;
```
same thing here with eslint-disable

alexzhang1618 commented on src/lib/api/UserAPI.ts:
```diff
@@ -41,3 +55,113 @@ export const getUserByHandle = async (token: string, handle: string): Promise<Pu
+ * @param user Profile changes
+ * @returns User's full profile
+ */
+export const updateCurrentUser = async (
```
Thoughts on renaming this function name to `updateCurrentUserProfile` or something? at a glance it looked like this function was switching which user was logged in

alexzhang1618 commented on src/lib/hoc/withAccessType.ts:
```diff
@@ -51,19 +51,21 @@ export default function withAccessType(
-    // If no user cookie, try to re-generate one
-    if (!userCookie || !userAccessLevel) {
-      try {
+    try {
+      user = JSON.parse(userCookie);
+      userAccessLevel = user.accessType;
+      // If no user cookie, try to re-generate one
+      if (!userCookie || !userAccessLevel) {
```
seeking to learn: what does this change?

SheepTester replied:
This is to prevent locking the user out if their user cookie contains malformed JSON for some reason

I made this change because my initial approach to update the user cookie after changing the profile was to set it to an empty string to force the server to regenerate it

alexzhang1618 commented on src/pages/profile/edit.tsx:
```diff
@@ -0,0 +1,660 @@
+                +graduationYear !== user.graduationYear ||
+                bio !== user.bio ||
+                isAttendancePublic !== user.isAttendancePublic ||
+                newPassword.length > 0
```
Can this be bundled with `hasChange`? (i.e.) `hasChange = hasProfileChange && hasSocialMediaChange`

SheepTester committed:
Merge remote-tracking branch 'origin/main' into sean/edit-profile

alexzhang1618 commented on src/pages/profile/edit.tsx:
```diff
@@ -0,0 +1,660 @@
+              let changed = false;
+              if (
+                firstName !== user.firstName ||
+                lastName !== user.lastName ||
```
Thoughts on putting this into a `handleSubmit` function to make the JSX section more concise?

alexzhang1618 commented on src/pages/profile/edit.tsx:
```diff
@@ -0,0 +1,660 @@
+                            if (file) {
+                              setPfp(file);
+                            }
+                          }}
```
Same question as in EventDetailsForm

alexzhang1618 reviewed (commented):
Functionality works great! Thanks for also refactoring the other modals, EventDetailsForm, and the Add to Apple Calendar button 😁 

Some questions/comments I had:

SheepTester committed:
address feedback

- remove redundant async onChange
- reduce code duplication with `hasChange`
- move `handleSubmit` out of JSX
- rename `updateCurrentUser` to `updateCurrentUserProfile` so it doesn't imply you're switching users
- hide heading collapsible arrows from screenreaders
- moved icons to SocialMediaIcon component

SheepTester committed:
apparently `bio` can be null

SheepTester committed:
Use TypeScript-safe no-unused-vars ESLint rule to fix false positives

alexzhang1618 reviewed (approved):
lgtm 🚢

farisashai commented on src/lib/api/UserAPI.ts:
```diff
@@ -1,9 +1,23 @@
 import { config } from '@/lib';
+import { UUID } from '@/lib/types';
```
```suggestion
import type { UUID } from '@/lib/types';
```

farisashai reviewed (changes requested):
Everything looks great and functional - couple UI nitpicks.  

- [x] I'm on mobile and a dropdown arrows shows up on both the left and right side of the details tag which seems unintended.

- [x] Thoughts on "Edit Profile" -> "Manage Account" since not all of the edits on this page are publicly facing. 

- [x] We have a route on the website `/logout` but no links to it anywhere so far. I think this page makes the most sense to include that link since it wouldn't be used too commonly by most users. 

- [x] I think we have something similar on some other form components but it would help a lot to highlight which fields have been edited and are currently unsaved so I can know before clicking the cancel or save button and making unintended edits.

SheepTester committed:
Update src/lib/api/UserAPI.ts

Co-authored-by: Faris Ashai <farisashai@gmail.com>

raymosun reviewed (changes requested):
Hey this looks great! The cropper modal is really well done. Just a few concerns: 

1. Is this delete button never available? Maybe we can remove it.
![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/30709485/95b805b7-deb6-44db-9e01-05ef3d22e0a1)

2. In dark mode, the placeholder color doesn't have enough contrast. (Also the character count color)
![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/30709485/bf56f3b2-0a98-417c-ba20-f8f2818dcca8)
![image](https://github.com/acmucsd/membership-portal-ui-v2/assets/30709485/1f7bcc9d-df9f-4c64-a5a3-b3bc06185b98)

3. ~Edit: I can't seem to see my current resume on the page~ Editedit: I deleted my cookie and it's fine, dunno what was up with that.

raymosun commented:
> Everything looks great and functional - couple UI nitpicks.
> 
> 1. I'm on mobile and a dropdown arrows shows up on both the left and right side of the details tag which seems unintended.
> 2. Thoughts on "Edit Profile" -> "Manage Account" since not all of the edits on this page are publicly facing.
> 3. We have a route on the website `/logout` but no links to it anywhere so far. I think this page makes the most sense to include that link since it wouldn't be used too commonly by most users.
> 4. I think we have something similar on some other form components but it would help a lot to highlight which fields have been edited and are currently unsaved so I can know before clicking the cancel or save button and making unintended edits.

Second Faris's fourth point about which fields have been edited. Right now it's a little confusing when changes are actually updated. Certain fields like profile picture and resume are updated at the moment you upload while other fields are only saved when you click save at the bottom. I think it's also a little confusing that pressing enter in one field will attempt to save all of them, including the password. Maybe we should think about moving the reset password to it's own section. Also, editing the email address is a pretty big change since it require's reverifying your email and I wonder if people can get locked out of their accounts by changing it, so maybe that field also needs more warning. Perhaps we can take inspiration from the v1 profile edit page where each section, separated by lines, gets saved together.

However, I think we should go ahead and merge this branch and then address those issues in a future PR.

SheepTester commented:
> 3\. ~Edit: I can't seem to see my current resume on the page~ Editedit: I deleted my cookie and it's fine, dunno what was up with that.

Your cookie caches your user data locally, so it can get stale when you make changes to your profile (such as earning points) from a different browser or device. We probably should look into better ways to store the user object

SheepTester committed:
address some feedback:
- make placeholder/char count more visible in dark mode
- actually update the user's resume visibility preference
- update edit profile -> manage account
- position sticky the save button when there are unsaved changes
- made dark-text-on-bg-2 (the placeholder color) distinct from -bg-1

removing overflow: auto on main was to allow for position: sticky. it seems to have come from the leaderboard PR so whatever

SheepTester committed:
Merge branch 'sean/edit-profile' of https://github.com/acmucsd/membership-portal-ui-v2 into sean/edit-profile

SheepTester committed:
Merge remote-tracking branch 'origin/main' into sean/edit-profile

SheepTester committed:
Indicate changed fields, add logout btn, and:
- fix iOS Safari double arrow
- fix select being unreadable (white on white) in dark mode
- ensure `user` object is up to date

SheepTester committed:
sanity check if cover image is too large

farisashai reviewed (approved):
I think the reset password form should be separate visually and with its own action button since it's destructive and not the same as updating profile fields. aside from that, this should be fine to merge and iterate later
